File: python/Problem53.py
# Problem: 53. Maximum Subarray
# URL: https://leetcode.com/problems/maximum-subarray/
# Difficulty: Medium
# Tags: Array

from typing import List

class Solution:
  def maxSubArray(self, nums: List[int]) -> int:
    sum: int = 0
    maxSum: int = nums[0]
    for num in nums:
      sum += num
      if sum > maxSum:
        maxSum = sum
      if sum < 0:
        sum = 0
    return maxSum

File: cpp/Problem53.cpp
/*
 * Problem: 53. Maximum Subarray
 * URL: https://leetcode.com/problems/maximum-subarray/
 * Difficulty: Medium
 * Tags: Array
 */

#include <vector>

using namespace std;

class Solution {
public:
	int maxSubArray(vector<int>& nums) {
		int sum = 0, maxSum = nums[0];
		for (int i = 0, size = nums.size(); i < size; i++) {
			sum += nums[i];
			if (sum > maxSum) {
				maxSum = sum;
			}
			if (sum < 0) {
				sum = 0;
			}
		}
		return maxSum;
	}
};